## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#687be8e1-0948-44f4-bf81-4a5a77d860e4) |
| mint-compression-fs.sh | :x: [more...](#afbcfee3-1e80-4090-956d-cd4a52b56900) |
| mint-fs.sh | :x: [more...](#93455837-b6b5-4276-88dd-22c47b42a337) |
<a name="93455837-b6b5-4276-88dd-22c47b42a337"></a>
### 8283-7bb83e2/mint-fs.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.52:30387
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 437868c0bf9d:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 0 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 42 seconds
(4/14) Running aws-sdk-ruby tests ... done in 2 seconds
(5/14) Running awscli tests ... done in 1 minutes and 50 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 39 seconds
(8/14) Running minio-dotnet tests ... done in 26 seconds
(9/14) Running minio-go tests ... done in 55 seconds
(10/14) Running minio-java tests ... done in 2 minutes and 1 seconds
(11/14) Running minio-js tests ... done in 46 seconds
(12/14) Running minio-py tests ... FAILED in 3 minutes and 38 seconds
{
  "function": "presigned_get_object(bucket_name, object_name, expires, response_headers, request_date)",
  "message": "AccessDenied: message: Access Denied",
  "status": "FAIL",
  "duration": 106,
  "args": {
    "object_name": "d26a39aa-9892-4881-9ae8-c225c0d6818d",
    "response.status": 403,
    "response.headers": "{\"_container\": {\"accept-ranges\": [\"Accept-Ranges\", \"bytes\"], \"content-length\": [\"Content-Length\", \"440\"], \"content-security-policy\": [\"Content-Security-Policy\", \"block-all-mixed-content\"], \"content-type\": [\"Content-Type\", \"application/xml\"], \"server\": [\"Server\", \"MinIO/DEVELOPMENT.2019-09-21T00-47-17Z\"], \"vary\": [\"Vary\", \"Origin\"], \"x-amz-request-id\": [\"X-Amz-Request-Id\", \"15C64EE673E549CE\"], \"x-xss-protection\": [\"X-Xss-Protection\", \"1; mode=block\"], \"date\": [\"Date\", \"Sat, 21 Sep 2019 01:11:11 GMT\"]}}",
    "response.reason": "Forbidden",
    "response._body": "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Error><Code>AccessDenied</Code><Message>Request has expired</Message><Key>d26a39aa-9892-4881-9ae8-c225c0d6818d</Key><BucketName>minio-py-test-870e25fc-4ffd-46e2-8784-d20e5b84abbe</BucketName><Resource>/minio-py-test-870e25fc-4ffd-46e2-8784-d20e5b84abbe/d26a39aa-9892-4881-9ae8-c225c0d6818d</Resource><RequestId>15C64EE673E549CE</RequestId><HostId>2357f4a7-e2c7-4901-9af1-88ea92e69c28</HostId></Error>",
    "bucket_name": "minio-py-test-870e25fc-4ffd-46e2-8784-d20e5b84abbe",
    "presigned_get_object_url": "http://72.28.97.52:30387/minio-py-test-870e25fc-4ffd-46e2-8784-d20e5b84abbe/d26a39aa-9892-4881-9ae8-c225c0d6818d?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=minio%2F20190921%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20190921T011106Z&X-Amz-Expires=5&X-Amz-SignedHeaders=host&X-Amz-Signature=ba7af58420e800f62eafa67589b4cf65f857d372f4b7b0231a4dcbd99285dd27"
  },
  "name": "minio-py:test_presigned_get_object_expiry_5sec",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1341, in test_presigned_get_object_expiry_5sec\n    object_name).get_exception()\nminio.error.AccessDenied: AccessDenied: message: Access Denied\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 2036, in main\n    test_presigned_get_object_expiry_5sec(client, log_output)\n  File \"/mint/run/core/minio-py/tests.py\", line 1355, in test_presigned_get_object_expiry_5sec\n    raise Exception(err)\nException: AccessDenied: message: Access Denied\n"
}

Executed 11 out of 14 tests successfully.
```
<a name="afbcfee3-1e80-4090-956d-cd4a52b56900"></a>
### 8283-7bb83e2/mint-compression-fs.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.53:31891
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 8e7ed7e6bcac:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 0 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 42 seconds
(4/14) Running aws-sdk-ruby tests ... done in 2 seconds
(5/14) Running awscli tests ... done in 1 minutes and 53 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 41 seconds
(8/14) Running minio-dotnet tests ... done in 24 seconds
(9/14) Running minio-go tests ... done in 57 seconds
(10/14) Running minio-java tests ... done in 2 minutes and 3 seconds
(11/14) Running minio-js tests ... done in 45 seconds
(12/14) Running minio-py tests ... FAILED in 3 minutes and 42 seconds
{
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1341, in test_presigned_get_object_expiry_5sec\n    object_name).get_exception()\nminio.error.AccessDenied: AccessDenied: message: Access Denied\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 2036, in main\n    test_presigned_get_object_expiry_5sec(client, log_output)\n  File \"/mint/run/core/minio-py/tests.py\", line 1355, in test_presigned_get_object_expiry_5sec\n    raise Exception(err)\nException: AccessDenied: message: Access Denied\n",
  "status": "FAIL",
  "duration": 62,
  "args": {
    "response.status": 403,
    "bucket_name": "minio-py-test-a5e6b07a-3c01-44b2-afb1-f319e5912e76",
    "response._body": "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Error><Code>AccessDenied</Code><Message>Request has expired</Message><Key>2c06f9f3-19e1-49ff-b2b7-75478d2bc3ae</Key><BucketName>minio-py-test-a5e6b07a-3c01-44b2-afb1-f319e5912e76</BucketName><Resource>/minio-py-test-a5e6b07a-3c01-44b2-afb1-f319e5912e76/2c06f9f3-19e1-49ff-b2b7-75478d2bc3ae</Resource><RequestId>15C64EE8E334578B</RequestId><HostId>0fc5b432-ab87-44f7-b518-998db54f10fe</HostId></Error>",
    "response.headers": "{\"_container\": {\"accept-ranges\": [\"Accept-Ranges\", \"bytes\"], \"content-length\": [\"Content-Length\", \"440\"], \"content-security-policy\": [\"Content-Security-Policy\", \"block-all-mixed-content\"], \"content-type\": [\"Content-Type\", \"application/xml\"], \"server\": [\"Server\", \"MinIO/DEVELOPMENT.2019-09-21T00-47-17Z\"], \"vary\": [\"Vary\", \"Origin\"], \"x-amz-request-id\": [\"X-Amz-Request-Id\", \"15C64EE8E334578B\"], \"x-xss-protection\": [\"X-Xss-Protection\", \"1; mode=block\"], \"date\": [\"Date\", \"Sat, 21 Sep 2019 01:11:22 GMT\"]}}",
    "object_name": "2c06f9f3-19e1-49ff-b2b7-75478d2bc3ae",
    "presigned_get_object_url": "http://72.28.97.53:31891/minio-py-test-a5e6b07a-3c01-44b2-afb1-f319e5912e76/2c06f9f3-19e1-49ff-b2b7-75478d2bc3ae?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=minio%2F20190921%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20190921T011116Z&X-Amz-Expires=5&X-Amz-SignedHeaders=host&X-Amz-Signature=f1dc1f6d6293d77476c3a4e24435c3abf01e8a0bf1d49691b545afe4c8492eee",
    "response.reason": "Forbidden"
  },
  "name": "minio-py:test_presigned_get_object_expiry_5sec",
  "message": "AccessDenied: message: Access Denied",
  "function": "presigned_get_object(bucket_name, object_name, expires, response_headers, request_date)"
}

Executed 11 out of 14 tests successfully.
```
<a name="687be8e1-0948-44f4-bf81-4a5a77d860e4"></a>
### 8283-7bb83e2/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.55:30020
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp f624964d79a4:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 3 seconds
(2/14) Running aws-sdk-java tests ... done in 2 seconds
(3/14) Running aws-sdk-php tests ... done in 49 seconds
(4/14) Running aws-sdk-ruby tests ... done in 14 seconds
(5/14) Running awscli tests ... done in 2 minutes and 27 seconds
(6/14) Running healthcheck tests ... done in 1 seconds
(7/14) Running mc tests ... done in 1 minutes and 3 seconds
(8/14) Running minio-dotnet tests ... done in 1 minutes and 26 seconds
(9/14) Running minio-go tests ... FAILED in 3 minutes and 38 seconds
{
  "args": {},
  "duration": 68221,
  "error": "The specified key does not exist.",
  "function": "testStorageClassMetadataCopyObject()",
  "message": "Stat failed",
  "name": "minio-go: testStorageClassMetadataCopyObject",
  "status": "FAIL"
}

Executed 8 out of 14 tests successfully.
```

--- a/cmd/http-tracer.go
+++ b/cmd/http-tracer.go
@@ -18,7 +18,6 @@ package cmd
 
 import (
 	"bytes"
-	"fmt"
 	"io"
 	"io/ioutil"
 	"net"
@@ -33,8 +32,6 @@ import (
 	"github.com/minio/minio/pkg/handlers"
 	trace "github.com/minio/minio/pkg/trace"
 )
-
-var traceBodyPlaceHolder = []byte("<BODY>")
 
 // recordRequest - records the first recLen bytes
 // of a given io.Reader
@@ -78,81 +75,7 @@ func (r *recordRequest) Data() []byte {
 		return r.buf.Bytes()
 	}
 	// ... otherwise we return <BODY> placeholder
-	return traceBodyPlaceHolder
-}
-
-// recordResponseWriter - records the first recLen bytes
-// of a given http.ResponseWriter
-type recordResponseWriter struct {
-	// Data source to record
-	http.ResponseWriter
-	// Response body should be logged
-	logBody bool
-	// Internal recording buffer
-	headers bytes.Buffer
-	body    bytes.Buffer
-	// The status code of the current HTTP request
-	statusCode int
-	// Indicate if headers are written in the log
-	headersLogged bool
-	// number of bytes written
-	bytesWritten int
-}
-
-// Write the headers into the given buffer
-func (r *recordResponseWriter) writeHeaders(w io.Writer, statusCode int, headers http.Header) {
-	n, _ := fmt.Fprintf(w, "%d %s\n", statusCode, http.StatusText(statusCode))
-	r.bytesWritten += n
-	for k, v := range headers {
-		n, _ := fmt.Fprintf(w, "%s: %s\n", k, v[0])
-		r.bytesWritten += n
-	}
-}
-
-// Record the headers.
-func (r *recordResponseWriter) WriteHeader(i int) {
-	r.statusCode = i
-	if !r.headersLogged {
-		r.writeHeaders(&r.headers, i, r.ResponseWriter.Header())
-		r.headersLogged = true
-	}
-	r.ResponseWriter.WriteHeader(i)
-}
-
-func (r *recordResponseWriter) Write(p []byte) (n int, err error) {
-	n, err = r.ResponseWriter.Write(p)
-	r.bytesWritten += n
-	if !r.headersLogged {
-		// We assume the response code to be '200 OK' when WriteHeader() is not called,
-		// that way following Golang HTTP response behavior.
-		r.writeHeaders(&r.headers, http.StatusOK, r.ResponseWriter.Header())
-		r.headersLogged = true
-	}
-	if r.statusCode >= http.StatusBadRequest || r.logBody {
-		// Always logging error responses.
-		r.body.Write(p)
-	}
-	return n, err
-}
-
-func (r *recordResponseWriter) Size() int {
-	return r.bytesWritten
-}
-
-// Calls the underlying Flush.
-func (r *recordResponseWriter) Flush() {
-	r.ResponseWriter.(http.Flusher).Flush()
-}
-
-// Return response body.
-func (r *recordResponseWriter) Body() []byte {
-	// If there was an error response or body logging is enabled
-	// then we return the body contents
-	if r.statusCode >= http.StatusBadRequest || r.logBody {
-		return r.body.Bytes()
-	}
-	// ... otherwise we return the <BODY> place holder
-	return traceBodyPlaceHolder
+	return logger.BodyPlaceHolder
 }
 
 // getOpName sanitizes the operation name for mc
@@ -205,16 +128,15 @@ func Trace(f http.HandlerFunc, logBody bool, w http.ResponseWriter, r *http.Requ
 		Headers:  reqHeaders,
 		Body:     reqBodyRecorder.Data(),
 	}
-
-	// Setup a http response body recorder
-	respBodyRecorder := &recordResponseWriter{ResponseWriter: w, logBody: logBody}
-	f(logger.NewResponseWriter(respBodyRecorder), r)
+	rw, _ := w.(*logger.ResponseWriter)
+	rw.LogBody = logBody
+	f(rw, r)
 
 	rs := trace.ResponseInfo{
 		Time:       time.Now().UTC(),
-		Headers:    respBodyRecorder.Header().Clone(),
-		StatusCode: respBodyRecorder.statusCode,
-		Body:       respBodyRecorder.Body(),
+		Headers:    rw.Header().Clone(),
+		StatusCode: rw.StatusCode,
+		Body:       rw.Body(),
 	}
 
 	if rs.StatusCode == 0 {
@@ -225,9 +147,10 @@ func Trace(f http.HandlerFunc, logBody bool, w http.ResponseWriter, r *http.Requ
 	t.RespInfo = rs
 
 	t.CallStats = trace.CallStats{
-		Latency:     rs.Time.Sub(rq.Time),
-		InputBytes:  reqBodyRecorder.Size(),
-		OutputBytes: respBodyRecorder.Size(),
+		Latency:         rs.Time.Sub(rw.StartTime),
+		InputBytes:      reqBodyRecorder.Size(),
+		OutputBytes:     rw.Size(),
+		TimeToFirstByte: rw.TimeToFirstByte,
 	}
 	return t
 }

--- a/cmd/logger/audit.go
+++ b/cmd/logger/audit.go
@@ -1,5 +1,5 @@
 /*
- * MinIO Cloud Storage, (C) 2018 MinIO, Inc.
+ * MinIO Cloud Storage, (C) 2018, 2019 MinIO, Inc.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -17,6 +17,9 @@
 package logger
 
 import (
+	"bytes"
+	"fmt"
+	"io"
 	"net/http"
 	"time"
 
@@ -27,9 +30,18 @@ import (
 // ResponseWriter - is a wrapper to trap the http response status code.
 type ResponseWriter struct {
 	http.ResponseWriter
-	statusCode      int
-	startTime       time.Time
-	timeToFirstByte time.Duration
+	StatusCode int
+	// Response body should be logged
+	LogBody         bool
+	TimeToFirstByte time.Duration
+	StartTime       time.Time
+	// number of bytes written
+	bytesWritten int
+	// Internal recording buffer
+	headers bytes.Buffer
+	body    bytes.Buffer
+	// Indicate if headers are written in the log
+	headersLogged bool
 }
 
 // NewResponseWriter - returns a wrapped response writer to trap
@@ -37,31 +49,75 @@ type ResponseWriter struct {
 func NewResponseWriter(w http.ResponseWriter) *ResponseWriter {
 	return &ResponseWriter{
 		ResponseWriter: w,
-		statusCode:     http.StatusOK,
-		startTime:      time.Now().UTC(),
+		StatusCode:     http.StatusOK,
+		StartTime:      time.Now().UTC(),
 	}
 }
 
 func (lrw *ResponseWriter) Write(p []byte) (int, error) {
 	n, err := lrw.ResponseWriter.Write(p)
+	lrw.bytesWritten += n
+	if lrw.TimeToFirstByte == 0 {
+		lrw.TimeToFirstByte = time.Now().UTC().Sub(lrw.StartTime)
+	}
+	if !lrw.headersLogged {
+		// We assume the response code to be '200 OK' when WriteHeader() is not called,
+		// that way following Golang HTTP response behavior.
+		lrw.writeHeaders(&lrw.headers, http.StatusOK, lrw.Header())
+		lrw.headersLogged = true
+	}
+	if lrw.StatusCode >= http.StatusBadRequest || lrw.LogBody {
+		// Always logging error responses.
+		lrw.body.Write(p)
+	}
 	if err != nil {
 		return n, err
-	}
-	if lrw.timeToFirstByte == 0 {
-		lrw.timeToFirstByte = time.Now().UTC().Sub(lrw.startTime)
 	}
 	return n, err
 }
 
+// Write the headers into the given buffer
+func (lrw *ResponseWriter) writeHeaders(w io.Writer, statusCode int, headers http.Header) {
+	n, _ := fmt.Fprintf(w, "%d %s\n", statusCode, http.StatusText(statusCode))
+	lrw.bytesWritten += n
+	for k, v := range headers {
+		n, _ := fmt.Fprintf(w, "%s: %s\n", k, v[0])
+		lrw.bytesWritten += n
+	}
+}
+
+// BodyPlaceHolder returns a dummy body placeholder
+var BodyPlaceHolder = []byte("<BODY>")
+
+// Body - Return response body.
+func (lrw *ResponseWriter) Body() []byte {
+	// If there was an error response or body logging is enabled
+	// then we return the body contents
+	if lrw.StatusCode >= http.StatusBadRequest || lrw.LogBody {
+		return lrw.body.Bytes()
+	}
+	// ... otherwise we return the <BODY> place holder
+	return BodyPlaceHolder
+}
+
 // WriteHeader - writes http status code
 func (lrw *ResponseWriter) WriteHeader(code int) {
-	lrw.statusCode = code
+	lrw.StatusCode = code
+	if !lrw.headersLogged {
+		lrw.writeHeaders(&lrw.headers, code, lrw.ResponseWriter.Header())
+		lrw.headersLogged = true
+	}
 	lrw.ResponseWriter.WriteHeader(code)
 }
 
 // Flush - Calls the underlying Flush.
 func (lrw *ResponseWriter) Flush() {
 	lrw.ResponseWriter.(http.Flusher).Flush()
+}
+
+// Size - reutrns the number of bytes written
+func (lrw *ResponseWriter) Size() int {
+	return lrw.bytesWritten
 }
 
 // AuditTargets is the list of enabled audit loggers
@@ -80,9 +136,9 @@ func AuditLog(w http.ResponseWriter, r *http.Request, api string, reqClaims map[
 	var timeToFirstByte time.Duration
 	lrw, ok := w.(*ResponseWriter)
 	if ok {
-		statusCode = lrw.statusCode
-		timeToResponse = time.Now().UTC().Sub(lrw.startTime)
-		timeToFirstByte = lrw.timeToFirstByte
+		statusCode = lrw.StatusCode
+		timeToResponse = time.Now().UTC().Sub(lrw.StartTime)
+		timeToFirstByte = lrw.TimeToFirstByte
 	}
 
 	vars := mux.Vars(r)

--- a/pkg/trace/trace.go
+++ b/pkg/trace/trace.go
@@ -33,9 +33,10 @@ type Info struct {
 
 // CallStats records request stats
 type CallStats struct {
-	InputBytes  int           `json:"inputbytes"`
-	OutputBytes int           `json:"outputbytes"`
-	Latency     time.Duration `json:"latency"`
+	InputBytes      int           `json:"inputbytes"`
+	OutputBytes     int           `json:"outputbytes"`
+	Latency         time.Duration `json:"latency"`
+	TimeToFirstByte time.Duration `json:"timetofirstbyte"`
 }
 
 // RequestInfo represents trace of http request

